src/Symfony/Component/HttpFoundation/StreamedJsonResponse.php

```diff
@@ -0,0 +1,177 @@
+ * efficient resource saving JSON response.
+ *
+ * It uses flush after a specified flush size to directly stream the data.
+ *
+ * @see flush()
+ *
+ * @author Alexander Schranz <alexander@sulu.io>
+ *
+ * Example usage:
+ *
+ * function loadArticles(): \Generator
+ *     // some streamed loading
+ *     yield ['title' => 'Article 1'];
+ *     yield ['title' => 'Article 2'];
+ *     yield ['title' => 'Article 3'];
+ * }),
+ *
+ * $response = new StreamedJsonResponse(
+ *     // json structure with generators in which will be streamed
+ *     [
+ *         '_embedded' => [
+ *             'articles' => loadArticles(), // any generator which you want to stream as list of data
+ *         ],
+ *     ],
+ * );
+ */
+class StreamedJsonResponse extends StreamedResponse
+{
+    private const PLACEHOLDER = '__symfony_json__';
+
+    private int $encodingOptions = JsonResponse::DEFAULT_ENCODING_OPTIONS;
+
+    /**
+     * @param mixed[]                        $data      JSON Data containing PHP generators which will be streamed as list of data
+     * @param int                            $status    The HTTP status code (200 "OK" by default)
+     * @param array<string, string|string[]> $headers   An array of HTTP headers
+     * @param int                            $flushSize After every which item of a generator the flush function should be called
+     */
+    public function __construct(
+        private readonly array $data,
+        int $status = 200,
+        array $headers = [],
+        private int $flushSize = 100,
+    ) {
+        parent::__construct($this->stream(...), $status, $headers);
+
+        if (!$this->headers->get('Content-Type')) {
+            $this->headers->set('Content-Type', 'application/json');
+        }
+    }
+
+    private function stream(): void
+    {
+        $generators = [];
+        $structure = $this->data;
+
+        array_walk_recursive($structure, function (&$item) use (&$generators) {
+            // generators should be used but for better DX all kind of Traversable are supported
+            if ($item instanceof \Traversable && !$item instanceof \JsonSerializable) {
+                $generators[] = $item;
+                $item = self::PLACEHOLDER;
+            } elseif (self::PLACEHOLDER === $item) {
+                // if the placeholder is already in the structure it should be replaced with a new one that explode
+                // works like expected for the structure
+                $generators[] = $item;
+            }
+        });
+
+        $jsonEncodingOptions = \JSON_THROW_ON_ERROR | $this->getEncodingOptions();
+        $keyEncodingOptions = $jsonEncodingOptions & ~\JSON_NUMERIC_CHECK;
+
+        $jsonParts = explode('"'.self::PLACEHOLDER.'"', json_encode($structure, $jsonEncodingOptions));
+
+        foreach ($generators as $index => $generator) {
+            // send first and between parts of the structure
+            echo $jsonParts[$index];
```

Comment:
are we 100% sure that `array_walk_recursive` and `json_encode` will iterate over the items in the exact same order (I'm thinking about (circular) references, for instance) and placeholder position will be synchronized with `$generators` keys?

Also, what if the string PLACEHOLDER is part of the serialized output? 
ie. `["__symfony_json__" => "foo"]` or `new SerializableEntity(name: '__symfony_json__')`, both case will generated `'"'.self::PLACEHOLDER.'"'` but won't trigger the `array_walk_recursive`

IMHO we should get rid of the second case by using a really random placeholder.

For the first issue, we could "labelize" the generator

```
$generators = [];

if ($item instanceof \Traversable && !$item instanceof \JsonSerializable) {
  $generators[] = $item;
  $item = self::PLACEHOLDER.count($generators);
}
```

latter
```
$jsonParts = preg_split('/"'.self::PLACEHOLDER.'(\d++)"/', json_encode($structure, $jsonEncodingOptions), -1, PREG_SPLIT_DELIM_CAPTURE);
for ($i=0; $i < count($jsonParts); i+=2) {
  echo $jsonParts[$i];
  if ($i + 1 < count($jsonParts)) {
    $generator = $generators[(int) $jsonParts[$i+1]];
  }
}
```

Reply:
A random placeholder does not guarantee that it won't be part of the data. It only makes it non-deterministic.
The case where the placeholder itself appear in the data is solved for values because `array_walk_recursive` injects a generator for such case, but it is indeed not solved for keys or JsonSerializable (for JsonSerializable, we could solve that by resolving them manually though)

Reply:
We could solve both by do the same as we are currently doing for values with `__symfony_json__`. I will have a look at it at the evening and create a test case for it.

Reply:
I fixed the handling, it is tested via `testResponseOtherTraversable` and `testPlaceholderAsKeyAndValueInStructure`.

Reply:
@jderusse Test added for it in `testResponseOtherTraversable` also tested when it contains key and value with the placeholder value.

Reply:
actually, I would expect this test to pass.

```
    public function testPlaceholderAsObjectStructure()
    {
        $object = new class() {
            public $__symfony_json__ = 'foo';
            public $bar = '__symfony_json__';
        };
        $content = $this->createSendResponse(
            [
                'object' => $object,
            ],
        );

        $this->assertSame('{"object":{"__symfony_json__":"foo","bar":"__symfony_json__"}}', $content);
    }
```

Reply:
also, this test breaks, which make me think that handling placeholder is really tricky


```
    public function testPlaceholderWithNested()
    {
        $content = $this->createSendResponse(
            [
                '__symfony_json__' => [
                    '__symfony_json__' => '__symfony_json__',
                ],
            ],
        );

        $this->assertSame('{"__symfony_json__":{"__symfony_json__":"__symfony_json__"}}', $content);
    }
```

Reply:
> actually, I would expect this test to pass.

@jderusse fixed that by handling all object via the placeholders.

> also, this test breaks, which make me think that handling placeholder is really tricky

Yeah not sure if we should optimize here for an edge case? this is something which we can not handle with `array_walk_recursive` currently.